substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrEvents.java

```diff
@@ -25,33 +25,49 @@
 package com.oracle.svm.jfr;
 
 import com.oracle.svm.core.annotate.Uninterruptible;
+import jdk.jfr.EventType;
+import jdk.jfr.internal.MetadataRepository;
+
+import java.util.List;
 
 /**
  * The event IDs depend on the metadata.xml and therefore vary between JDK versions.
  */
 public enum JfrEvents {
-    // TODO: we need to abstract the JDK version in some way.
-    // Event IDs should be fetched similar to how we do it in JfrTypes.
-    MetadataEvent(0),
-    CheckpointEvent(1),
-    ThreadStartEvent(255),
-    ThreadEndEvent(256),
-    DataLossEvent(335);
-
-    private final int id;
-
-    JfrEvents(int id) {
-        this.id = id;
+    ThreadStartEvent("jdk.ThreadStart"),
+    ThreadEndEvent("jdk.ThreadEnd"),
+    DataLossEvent("jdk.DataLoss");
+
+    private final long id;
+
+    JfrEvents(String name) {
+        this.id = getEventTypeId(name);
+    }
+
+    private static long getEventTypeId(String name) {
+        MetadataRepository metadata = MetadataRepository.getInstance();
+        List<EventType> eventTypes = metadata.getRegisteredEventTypes();
+        for (EventType eventType : eventTypes) {
+            if (name.equals(eventType.getName())) {
+                return eventType.getId();
+            }
+        }
+        return 0;
     }
 
     @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
-    public int getId() {
+    public long getId() {
         return id;
     }
 
     public static int getEventCount() {
-        // TODO: needs to return the count of all native events that are defined in the metadata.xml
-        // file. The highest id must match "eventCount - 1".
-        return 400;
+        MetadataRepository metadata = MetadataRepository.getInstance();
+        List<EventType> eventTypes = metadata.getRegisteredEventTypes();
```

Comment:
Does this include Java-level events as well (we only want the maximum id of the native events + 1).

Reply:
As far as I can tell, it includes everything that is defined in metadata.xml, I believe this is what you call 'native events'.

Reply:
Yes, as far as I know, everything in the `metadata.xml` file are the native events (or sometimes called JVM-level events). I had a look at `MetadataRepository.getRegisteredEventTypes()` and it includes both native events and `EventHandler`s for all registered Java-level events (i.e., classes that extend `jdk.jfr.Event`). So, it might be necessary to access `MetadataRepository.nativeEventTypes` instead.

Reply:
Hmm ok, bad. What would be the best realistic way to access nativeEventTypes, given that it is private? Good old reflection maybe? It's running on the host VM after all.

Reply:
Yes, reflection is probably the only way. At least, the field name seems to be stable across all JDK versions.

Reply:
I think I found a better way without reflection: Same procedure like how nativeEventTypes is constructed: scan all types from TypeLibrary.getTypes() and filter by PlatformEventType.